### PR TITLE
Add problems `chain` and `polygon`

### PR DESCRIPTION
**Pull request by tmigot**
@frapac I kept the existing docstring, but we should probably unify with the other problems, what do you think?

### CODE REVIEW

File: src/chain.jl [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+# Find the chain (of uniform density) of length L suspended between two points with minimal
+# potential energy.
+
+#   This is problem 4 in the COPS (Version 3) collection of 
+#   E. Dolan and J. More'
+#   see "Benchmarking Optimization Software with COPS"
+#   Argonne National Labs Technical Report ANL/MCS-246 (2004)
+
+function chain_model(args...; n::Int = default_nvar, kwargs...)
```
Comment (frapac) [inline src/chain.jl]: I would suggest to make the argument explicit here, as 
```julia
function chain_model(n::Int) 
```
what do you think?

File: src/chain.jl [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  @variable(nlp, x1[k = 1:(nh + 1)], start = 4 * abs(b - a) * k / nh * (1 / 2 * k / nh - tmin) + a)
+  @variable(
+    nlp,
+    x2[k = 1:(nh + 1)],
+    start =
+      (4 * abs(b - a) * k / nh * (1 / 2 * k / nh - tmin) + a) * (4 * abs(b - a) * (k / nh - tmin))
+  )
+  @variable(nlp, x3[k = 1:(nh + 1)], start = 4 * abs(b - a) * (k / nh - tmin))
+
+  @NLobjective(nlp, Min, x2[nh + 1])
```
Comment (frapac) [inline src/chain.jl]: We have to formulate the problem with `ScalarNonlinearFunction` in MOI to support ExaModels. As such, we have to formulate all nonlinear expressions with `@objective` and `@constraint`.

File: src/chain.jl [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  for j = 1:nh
+    @constraint(nlp, x1[j + 1] - x1[j] - 1 / 2 * h * (u[j] + u[j + 1]) == 0)
+  end
+  @constraint(nlp, x1[1] == a)
+  @constraint(nlp, x1[nh + 1] == b)
+  @constraint(nlp, x2[1] == 0)
+  @constraint(nlp, x3[1] == 0)
+  @constraint(nlp, x3[nh + 1] == L)
+
+  @NLconstraint(
```
Comment (frapac) [inline src/chain.jl]: Can we reformulate this expression with `@constraint` ?

File: src/chain.jl [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    nlp,
+    [j = 1:nh],
+    x2[j + 1] - x2[j] - 1 / 2 * h * (x1[j] * sqrt(1 + u[j]^2) + x1[j + 1] * sqrt(1 + u[j + 1]^2)) ==
+    0
+  )
+  @NLconstraint(
```
Comment (frapac) [inline src/chain.jl]: Can we reformulate this expression with `@constraint` ?

File: src/polygon.jl [hunk 5]
```diff
@@ -0,0 +1,29 @@
+# Find the polygon of maximal area, among polygons with nv sides and    
+#   diameter d <= 1
+
+#   This is problem 1 in the COPS (Version 3) collection of 
+#   E. Dolan and J. More'
+#   see "Benchmarking Optimization Software with COPS"
+#   Argonne National Labs Technical Report ANL/MCS-246 (2004)
+
+function polygon_model(args...; n::Int = default_nvar, kwargs...)
```
Comment (frapac) [inline src/polygon.jl]: As in the previous function, I would suggest making all the arguments explicit. How about using? 
```julia
function polygon_model(n::Int)
```

File: src/polygon.jl [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+  # impose an order to the angles
+  @constraint(nlp, θ[N] == π)
+  @constraint(nlp, r[N] == 0)
+  for i = 1:(N - 1)
+    @constraint(nlp, θ[i + 1] - θ[i] >= 0.0)
+  end
+  for i = 1:(N - 1)
+    for j = (i + 1):N
+      @NLconstraint(nlp, r[i]^2 + r[j]^2 - 2 * r[i] * r[j] * cos(θ[i] - θ[j]) - 1 <= 0)
```
Comment (frapac) [inline src/polygon.jl]: Can we reformulate this expression with `@constraint` ?

File: src/polygon.jl [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    end
+  end
+
+  @NLobjective(nlp, Min, -0.5 * sum(r[i] * r[i + 1] * sin(θ[i + 1] - θ[i]) for i = 1:(N - 1)))
```
Comment (frapac) [inline src/polygon.jl]: Can we reformulate this expression with `@objective` ?

**Review by frapac — CHANGES_REQUESTED** [review]
Thank you for your contribution @tmigot ! 
I only have a few suggestions, summarizing to:
- passing explicitly the arguments to the functions 
- using the newest NLP backend in JuMP (ie replacing `@NLobjective` by `@objective` and equivalently for `@NLconstraint`). This is required to support ExaModels.

The docstrings look good to me. Maybe add a line to specify that these two models are taken from `OptimizationProblems.jl`?

**Review by frapac — APPROVED** [review]
Looks good to me, thank you!

**Comment by tmigot** [comment]
@frapac I added the includes in the main file :). As mentioned in the report, polygon has several local minima so it is hard to test the objective value there.

**Pull request merged**